### PR TITLE
Examples of parallel training on one GPU using functorch.vmap with torchopt

### PR DESCRIPTION
**Pull request by Benjamin-eecs**
#23 

under `examples/functorch`  folder

cc @vmoens @waterhorse1 @zou3519 @Chillee

### CODE REVIEW

File: examples/functorch/parallel_train_torchopt.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        x = self.fc2(x)
+        x = F.log_softmax(x, -1)
+        return x
+
+
+class ParallelTrainFunctorchOriginal:
+    def __init__(self, loss_fn, lr):
+        self.loss_fn = loss_fn
+        self.lr = lr
+        self.func_model, _ = make_functional(MLPClassifier().to(DEVICE))
```
Comment (vmoens) [inline examples/functorch/parallel_train_torchopt.py:64]: Should a functional model have a device?

File: examples/functorch/parallel_train_torchopt.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    def init_fn(self, num_models):
+        models = [MLPClassifier().to(DEVICE) for _ in range(num_models)]
+        _, batched_weights, _ = combine_state_for_ensemble(models)
+        return batched_weights
+
+    def train_step_fn(self, weights, batch, targets):
+        def compute_loss(weights, batch, targets):
```
Comment (vmoens) [inline examples/functorch/parallel_train_torchopt.py:72]: Can't we make a separate method with this, rather than defining a fun in a method?

File: examples/functorch/parallel_train_torchopt.py (hunk 1 above)
Comment (vmoens) [reply examples/functorch/parallel_train_torchopt.py:64]: Also are you sure you want to use global variables, like DEVICE? IMO it can quickly become a problem (eg when working on distributed/mp settings etc)

**Review by vmoens — COMMENTED** [review]
Just a couple of comments here and there :)

File: examples/functorch/parallel_train_torchopt.py (hunk 1 above)
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:64]: Yes, I think you are right. I was just mimicking the way zou did in [functorch examples](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L75)

File: examples/functorch/parallel_train_torchopt.py (hunk 2 above)
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:72]: Same here, just mimicking the way zou did in functorch [parallel training examples](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L78-L93). Also, similar way to make it functional can be found in functorch [MAML omniglot example](https://github.com/pytorch/functorch/blob/main/examples/maml_omniglot/maml-omniglot-transforms.py#L125-L147).

Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:72]: I think in general Jax also, just [one example](https://huggingface.co/spaces/dalle-mini/dalle-mini/blob/53dade7fd869d7f66821a7179e9f508655b9afd0/tools/train/train.py#L553-L580) for demonstration. May be that's jax style :)

Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:72]: But I think loss_fn can be indeed defined outside train_step_fn.

**Comment by Benjamin-eecs** [comment]
> Just a couple of comments here and there :)

Also, there is a [colab](https://colab.research.google.com/drive/1YELlRebL3XLj6lsTceRPBaJwWE5WBVAs?usp=sharing) to play with :)

File: examples/functorch/parallel_train_torchopt.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        def compute_loss(weights, batch, targets):
+            output = self.func_model(weights, batch)
+            loss = self.loss_fn(output, targets)
+            return loss
+
+        grads, loss = grad_and_value(compute_loss)(weights, batch, targets)
+        # functional optimizer API is here now
+        updates, new_opt_state = optimizer.update(grads, opt_state, inplace=False)
+        new_weights = torchopt.apply_updates(weights, updates, inplace=False)
```
Comment (zou3519) [inline examples/functorch/parallel_train_torchopt.py:124]: Do all optimizers out there compute a quantity that is an update and apply the update in the final step? Are there any optimizers that do e.g. `param *= update` instead?

I've never come across one, but just curious

File: examples/functorch/parallel_train_torchopt.py (hunk 2 above)
Comment (zou3519) [reply examples/functorch/parallel_train_torchopt.py:72]: Either way works, we didn't think too much about code clarity when we wrote https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L78-L93

File: examples/functorch/parallel_train_torchopt.py (hunk 3 above)
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:124]: Yeah, me too, I once thought about that too! IMHO, I think the semantic meaning of ` updates` is the \delta `params`, which is the changes of the `params`. In the conventional sense of gradient-based optimization, `apply_updates` here does only add the `updates` onto `params`. But I am open to discussion, maybe in the larger scope of optimization. I will spend time looking into it.

File: examples/functorch/parallel_train_torchopt.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+import functorch
+import torchopt
+from functorch import combine_state_for_ensemble, grad_and_value, make_functional, vmap
+
+
+def make_spirals(n_samples, noise_std=0.0, rotations=1.0):
+    ts = torch.linspace(0, 1, n_samples, device=DEVICE)
```
Comment (XuehaiPan) [inline examples/functorch/parallel_train_torchopt.py:31]: `DEVICE` is a global variable defined in `if __name__ == '__main__'`. Please pass this as function argument instead.

File: examples/functorch/parallel_train_torchopt.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        x = self.fc2(x)
+        x = F.log_softmax(x, -1)
+        return x
+
+
+class ParallelTrainFunctorchOriginal:
+    def __init__(self, loss_fn, lr):
+        self.loss_fn = loss_fn
+        self.lr = lr
+        self.func_model, _ = make_functional(MLPClassifier().to(DEVICE))
```
Comment (XuehaiPan) [inline examples/functorch/parallel_train_torchopt.py:64]: Same as above, pass `device` as argument.

File: examples/functorch/parallel_train_torchopt.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+import argparse
+import math
+from collections import namedtuple
+from typing import Any, NamedTuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import functorch
```
Comment (XuehaiPan) [inline examples/functorch/parallel_train_torchopt.py:25]: Move this import above `import torch`.

**Review by XuehaiPan — CHANGES_REQUESTED** [review]
Please resolve the conversations and add a new line to the changelog.

File: examples/functorch/parallel_train_torchopt.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+import torchopt
+from functorch import combine_state_for_ensemble, grad_and_value, make_functional, vmap
```
[the remark this answers is not in the document]
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:27]: Please read [zou example](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L6) first, I try to make minimal changes to his example.

File: examples/functorch/parallel_train_torchopt.py (hunk 4 above)
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:31]: Please read [zou example](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L6).

File: examples/functorch/parallel_train_torchopt.py (hunk 5 above)
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:64]: Please read [zou example](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L6).

File: examples/functorch/parallel_train_torchopt.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            loss, batched_weights = parallel_train_step_fn(batched_weights, points, labels)
+            if i % 200 == 0:
+                print(loss)
+
+
+class ParallelTrainFunctorchTorchOpt:
+    def __init__(self, loss_fn, optimizer):
+        self.loss_fn = loss_fn
+        self.optimizer = optimizer
+        self.func_model, _ = make_functional(MLPClassifier().to(DEVICE))
```
[the remark this answers is not in the document]
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:106]: Please read [zou example](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L6).

File: examples/functorch/parallel_train_torchopt.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    )
+    args = parser.parse_args()
+
+    DEVICE = args.device
+    # Step 1: Make some spirals
+    points, labels = make_spirals(100, noise_std=0.05)
+    # Step 2: Define two-layer MLP and loss function
+    loss_fn = nn.NLLLoss()
+    # Step 3: Make the model functional(!!) and define a training function.
+    func_model, weights = make_functional(MLPClassifier().to(DEVICE))
```
[the remark this answers is not in the document]
Comment (Benjamin-eecs) [reply examples/functorch/parallel_train_torchopt.py:174]: To highlight the different implementations between original functorch example and our new example.

File: torchopt/_src/transform.py [hunk 10]
```diff
@@ -42,13 +42,23 @@
 
 
 ScaleState = base.EmptyState
+MaxInt32Value = torch.iinfo(torch.int32).max
```
Comment (XuehaiPan) [inline torchopt/_src/transform.py:45]:
```suggestion
INT32_MAX = torch.iinfo(torch.int32).max
```

File: torchopt/_src/transform.py [hunk 11]
```diff
@@ -42,13 +42,23 @@
...
+
+    Returns:
+        A counter incremeted by one, or max_int if the maximum precision is reached.
+    """
+    one = torch.ones(1, dtype=torch.int32, device=count[0].device)
 
     def f(c, g):
-        return c + 1 if g is not None else c
+        return (
+            c + (1 - torch.div(c, MaxInt32Value, rounding_mode='trunc')) * one
```
Comment (XuehaiPan) [inline torchopt/_src/transform.py:58]:
```suggestion
            c + (1 - torch.div(c, INT32_MAX, rounding_mode='trunc')) * one
```

**Review by XuehaiPan — APPROVED** [review]
